api_tests/lib/bitcoin.ts

```diff
@@ -0,0 +1,171 @@
+}
+
+interface Utxo {
+    // TODO: declare transactionId type
+    // Best to have specific transactionId type than using generic strings
+    txId: string;
+    value: number;
+    vout: number;
+}
+
+let _bitcoinRpcClient: BitcoinRpcClient;
+let _bitcoinConfig: BitcoinConfig;
+
+function createBitcoinRpcClient(btcConfig: BitcoinConfig) {
+    if (!btcConfig && !_bitcoinConfig) {
+        throw new Error("bitcoin configuration is needed");
+    }
+
+    if (!_bitcoinRpcClient || btcConfig !== _bitcoinConfig) {
+        _bitcoinRpcClient = new BitcoinRpcClient({
+            network: "regtest",
+            port: btcConfig.rpc_port,
+            host: btcConfig.rpc_host,
+            username: btcConfig.rpc_username,
+            password: btcConfig.rpc_password,
+        });
+        _bitcoinConfig = btcConfig;
+    }
+    return _bitcoinRpcClient;
+}
+
+module.exports.createClient = (btcConfig: BitcoinConfig) => {
+    return createBitcoinRpcClient(btcConfig);
+};
+
+module.exports.generate = async function(num: number = 1) {
+    return createBitcoinRpcClient(_bitcoinConfig).generate(num);
+};
+
+module.exports.activateSegwit = async function() {
+    const blockHeight = await createBitcoinRpcClient(
+        _bitcoinConfig
+    ).getBlockCount();
+    if (blockHeight < 432) {
+        await createBitcoinRpcClient(_bitcoinConfig).generate(432);
+    }
+};
+
+async function getFirstUtxoValueTransferredTo(txId: string, address: string) {
+    let satoshi = 0;
+    let tx = await _bitcoinRpcClient.getRawTransaction(txId, true);
+    let vout = tx.vout[0];
+
+    if (
+        vout.scriptPubKey.addresses.length === 1 &&
+        vout.scriptPubKey.addresses[0] === address
+    ) {
+        satoshi = sb.toSatoshi(vout.value);
+    }
+
+    return satoshi;
+}
+
+module.exports.get_first_utxo_value_transferred_to = getFirstUtxoValueTransferredTo;
+
+export class BitcoinWallet {
+    keypair: ECPair;
+    bitcoinUtxos: Utxo[];
+    _identity: {
+        address: string;
+        hash: Buffer;
+        output: Buffer;
+        pubkey: Buffer;
+        signature: Buffer;
+        input: Buffer;
+        witness: Buffer[];
+    };
+
+    constructor() {
+        this.keypair = ECPair.makeRandom({ rng: util.test_rng });
+        // TODO: Use wallet instead of array to track Bitcoin UTXOs
+        this.bitcoinUtxos = [];
+        this._identity = payments.p2wpkh({
+            pubkey: this.keypair.publicKey,
+            network: regtest,
+        });
+    }
+
+    identity() {
+        return this._identity;
+    }
+
+    async fund(btcValue: number) {
+        let txId = await _bitcoinRpcClient.sendToAddress(
+            this.identity().address,
+            btcValue
+        );
+        let raw_transaction = await _bitcoinRpcClient.getRawTransaction(txId);
+        let transaction = Transaction.fromHex(raw_transaction);
+        let entries: Out[] = transaction.outs;
+        for (let i in entries) {
```

Comment:
🙃 You could make this more sexy  
`this.bitcoinUtxos. concat(entries.filter....map..)`

Reply:
If you were to do this, you would have to overwrite `this.bitcoinUtxos` because `concat` [does not override the original arrays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat#Syntax)

Reply:
Can you provide the full gist of what you are suggesting?

Reply:
I just tried this and this seems to be working :)
```
this.bitcoinUtxos.push(
            ...entries
                .filter(entry => entry.script.equals(this.identity().output))
                .map((entry, index) => {
                    return {
                        txId: txId,
                        vout: index,
                        value: entry.value,
                    };
                })
        );
```

Reply:
Are you sure it works? did you run `cargo make api`

```
RpcError: bad-txns-in-belowout, value in (2.499963) < value out (9.999975) (code 16)
```